File: firebase/functions/lib/events/live_meetings/kick_participant.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import 'package:functions/utils/utils.dart';
import '../../on_call_function.dart';
import 'agora_api.dart';
import '../../utils/infra/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/events/live_meetings/live_meeting.dart';
import 'package:data_models/community/membership.dart';

class KickParticipant extends OnCallMethod<KickParticipantRequest> {
  AgoraUtils agoraUtils;
  KickParticipant({AgoraUtils? agoraUtils})
      : agoraUtils = agoraUtils ?? AgoraUtils(),
        super(
          'KickParticipant',
          (json) => KickParticipantRequest.fromJson(json),
        );

  @override
  Future<String> action(
    KickParticipantRequest request,
    CallableContext context,
  ) async {
    final eventPath = request.eventPath;

    await firestore.runTransaction((transaction) async {
      final event = await firestoreUtils.getFirestoreObject(
        transaction: transaction,
        path: eventPath,
        constructor: (map) => Event.fromJson(map),
      );

      final communityMembershipDoc = await firestore
          .document(
            'memberships/${context.authUid}/community-membership/${event.communityId}',
          )
          .get();

      final membership = Membership.fromJson(
        firestoreUtils.fromFirestoreJson(communityMembershipDoc.data.toMap()),
      );

      if (event.creatorId != context.authUid && !membership.isFacilitator) {
        throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
      }

      orElseUnauthorized(
        request.userToKickId != event.creatorId,
        logMessage: 'Event creator cannot be kicked from event.',
      );

      final liveMeeting = await firestoreUtils.getFirestoreObject(
        transaction: transaction,
        path: '$eventPath/live-meetings/${event.id}',
        constructor: (map) => LiveMeeting.fromJson(map),
      );

      // Kick participant
      final roomId = request.breakoutRoomId ?? liveMeeting.meetingId ?? '';
      await agoraUtils.kickParticipant(
        roomId: roomId,
        userId: request.userToKickId,
      );
    });

    return '';
  }
}

File: firebase/functions/src/events/live_meetings/kick_participant.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { orElseUnauthorized } from '../../utils/utils';
import { AgoraUtils } from './agora_api';
import { Event, Membership, LiveMeeting, membershipIsFacilitator } from '../../types';

interface KickParticipantRequest {
  eventPath: string;
  userToKickId: string;
  breakoutRoomId?: string;
}

export class KickParticipant extends OnCallMethod<KickParticipantRequest> {
  private agoraUtils: AgoraUtils;

  constructor(agoraUtils?: AgoraUtils) {
    super('KickParticipant', (json) => json as KickParticipantRequest);
    this.agoraUtils = agoraUtils ?? new AgoraUtils();
  }

  async action(request: KickParticipantRequest, context: functions.https.CallableContext): Promise<string> {
    const eventPath = request.eventPath;

    await firestore.runTransaction(async (transaction) => {
      const event: Event = await firestoreUtils.getFirestoreObject({
        transaction,
        path: eventPath,
        constructor: (map) => map as unknown as Event,
      });

      const communityMembershipSnap = await firestore
        .doc(`memberships/${context.auth?.uid}/community-membership/${(event as any).communityId}`)
        .get();
      const membership = firestoreUtils.fromFirestoreJson(communityMembershipSnap.data() ?? {}) as unknown as Membership;

      if ((event as any).creatorId !== context.auth?.uid && !membershipIsFacilitator((membership as any).status)) {
        throw new functions.https.HttpsError('failed-precondition', 'unauthorized');
      }

      orElseUnauthorized(
        request.userToKickId !== (event as any).creatorId,
        { logMessage: 'Event creator cannot be kicked from event.' }
      );

      const liveMeeting: LiveMeeting = await firestoreUtils.getFirestoreObject({
        transaction,
        path: `${eventPath}/live-meetings/${(event as any).id}`,
        constructor: (map) => map as unknown as LiveMeeting,
      });

      const roomId = request.breakoutRoomId ?? (liveMeeting as any).meetingId ?? '';
      await this.agoraUtils.kickParticipant({ roomId, userId: request.userToKickId });
    });

    return '';
  }
}
